sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroIO.java

```diff
@@ -760,6 +765,36 @@ private static void validateOutputComponent(String partialFilePattern) {
         + partialFilePattern);
   }
 
+  /**
+   * Retrieve the browse URL for a file pattern.
+   *
+   * @param validate Whether validation errors should cause an exception to throw.
+   * @return The browse URL, or null if the filePattern is invalid and validation is disabled.
+   */
+  @Nullable
+  private static String getBrowseUrl(String filePattern, boolean validate) {
+    if (!IOChannelUtils.hasFactory(filePattern)) {
+      // Browse URLs are only used for display data and shouldn't throw unless validation is
+      // enabled.
+      if (validate) {
+        throw new IllegalStateException(String.format("Invalid filePattern: %s", filePattern));
+      } else {
+        return null;
+      }
+    }
+
+    IOChannelFactory factory;
+    try {
+      factory = IOChannelUtils.getFactory(filePattern);
+    } catch (IOException e) {
+      // hasFactory checked above, should not throw
+      throw new AssertionError(String.format(
+          "Unexpected error while retrieving browse url for file pattern: %s", filePattern), e);
```

Comment:
If this sequence of calls is guaranteed to never throw, we should consider removing the checked exception and throwing an `IllegalArgumentException` if called improperly; If it can change from under you (which it probably shouldn't be able to), then this probably more of a `ConcurrentModificationException` or the like.

You don't need to make this change now, but maybe file a JIRA issue.